lib/pdk/module/release.rb

```diff
@@ -0,0 +1,260 @@
+require 'pdk'
+
+module PDK
+  module Module
+    class Release
+      def self.invoke(module_path, options = {})
+        new(module_path, options).run
+      end
+
+      attr_reader :options
+
+      attr_reader :module_path
+
+      def initialize(module_path, options = {})
+        @options = options
+
+        # TODO: Currently the release process can ONLY be run if the working directory IS the module root. However, in the future
+        # this WILL change, so we have the API arguments for it, but only accept `nil` for the first parameter
+        raise _('Running the release process outside of the working directory is not supported') unless module_path.nil?
+
+        if module_path.nil?
+          module_path = PDK::Util.module_root
+          raise PDK::CLI::ExitWithError, _('The module release process requires a valid module path') % { module_path: module_path } if module_path.nil?
+        end
+        raise PDK::CLI::ExitWithError, _('%{module_path} is not a valid module') % { module_path: module_path } unless PDK::Util.in_module_root?(module_path)
+        @module_path = module_path
+      end
+
+      def run
+        # Pre-release checks
+        unless force?
+          raise PDK::CLI::ExitWithError, _('The module is not PDK compatible') if requires_pdk_compatibility? && !pdk_compatible?
+          raise PDK::CLI::ExitWithError, _('The module is not Forge compatible') if requires_forge_compatibility? && !forge_compatible?
+        end
+
+        # Note that these checks are duplicated in the run_publish method, however it's a much better
+        # experience to fail early, than going through the whole process, only to error at the end knowing full well
+        # it'll fail anyway.
+        validate_publish_options!
+
+        run_validations(options) unless skip_validation?
+
+        PDK.logger.info _('Releasing %{module_name} - from version %{module_version}') % {
+          module_name:    module_metadata.data['name'],
+          module_version: module_metadata.data['version'],
+        }
+
+        PDK::Util::ChangelogGenerator.generate_changelog unless skip_changelog?
```

Comment:
If you move this below the conditional on lines 57-68, could you then remove the duplicate call at the end of that conditional (i.e. rather than generate the changelog and update the version number and then regenerate, just generate the changelog after possibly updating the version number)?

Reply:
Unfortunately we can't (right now).  One workflow is:

* Generate the changelog with the existing version number
* Compute the next version by parsing the changelog looking for Major, Minor, Patch changes 
* Bump the metadata with new version
* Regenerate the Changelog with the now "correct" version number

Unfortunately, due to the way the task is being called (via a Rake task in the module's context) I can't just generate the changelog "on the fly" (e.g. output to a temp file, or capture on STDOUT).  Ideally it would be:

* Generate the current changelog and output to STDOUT (don't save)
* Parse the changelog ....
...
* Generate the changelog and save to CHANGELOG.md

Reply:
If the changelog gem was in the PDK's context (not the modules), perhaps we'd have more control over it? and do smarter things.